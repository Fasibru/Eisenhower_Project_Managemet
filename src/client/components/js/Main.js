import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../css/Main.css';
import '../css/Task.css';
import Task from './Task';


class Main extends Component {
  // constructor(props) {
  //   super(props);
  // }
  static categorizeFilteredTasks(filteredTasks) {
    const categorizedFilteredTasks = {
      TasksCatA: [],
      TasksCatB: [],
      TasksCatC: [],
      TasksCatD: [],
    };

    // const { toggleEditTaskPopup } = this.props;

    filteredTasks.forEach((task) => {
      switch (task.category) {
        case 'A':
          categorizedFilteredTasks.TasksCatA.push(<Task key={task._id} className="task task-category-A" title={task.title} description={task.description} />);
          break;

        case 'B':
          categorizedFilteredTasks.TasksCatB.push(<Task key={task._id} className="task task-category-B" title={task.title} description={task.description} />);
          break;

        case 'C':
          categorizedFilteredTasks.TasksCatC.push(<Task key={task._id} className="task task-category-C" title={task.title} description={task.description} />);
          break;

        case 'D':
          categorizedFilteredTasks.TasksCatD.push(<Task key={task._id} className="task task-category-D" title={task.title} description={task.description} />);
          break;

        default:
          console.log(`There seems to be no category assigned to the following task: ${task}`);
      }
    });
    return categorizedFilteredTasks;
  }

  render() {
    const { filteredTasks } = this.props;
    const categorizedFilteredTasks = Main.categorizeFilteredTasks(filteredTasks);
    return (
      <main className="main">
        <div className="column-category">{categorizedFilteredTasks.TasksCatA}</div>
        <div className="column-category">{categorizedFilteredTasks.TasksCatB}</div>
        <div className="column-category">{categorizedFilteredTasks.TasksCatC}</div>
        <div className="column-category">{categorizedFilteredTasks.TasksCatD}</div>
      </main>
    );
  }
}

Main.propTypes = {
  filteredTasks: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default Main;
