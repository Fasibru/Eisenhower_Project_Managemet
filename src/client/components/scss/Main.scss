.main {
  grid-area: main;
  display: flex;
  background-color: #f5fbef;
  overflow-y: scroll;
}

.main__column-category {
  flex: 1;
  border-right: 1px solid black;
}

.column-category:last-child {
  border: none;
}