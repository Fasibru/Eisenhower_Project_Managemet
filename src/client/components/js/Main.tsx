/* eslint-disable no-undef */
import * as React from 'react';
// import PropTypes from 'prop-types';

import '../scss/Main.scss';
import '../scss/Task.scss';

import Task from './Task';

import { TaskType } from '../../../types/storeTypes';

interface TasksCatProps {
  className: string;
  key: string;
  task: TaskType;
}

interface CategorizeFilteredTasks {
  TasksCatA: Array<React.ReactElement<TasksCatProps>>;
  TasksCatB: Array<React.ReactElement<TasksCatProps>>;
  TasksCatC: Array<React.ReactElement<TasksCatProps>>;
  TasksCatD: Array<React.ReactElement<TasksCatProps>>;
}

interface MainProps {
  filteredTasks: TaskType[];
}

const categorizeFilteredTasks = (filteredTasks: TaskType[]) => {
  const categorizedFilteredTasks: CategorizeFilteredTasks = {
    TasksCatA: [],
    TasksCatB: [],
    TasksCatC: [],
    TasksCatD: [],
  };

  // sort filteredTasks array based on rank
  // filteredTasks.sort((a, b) => {
  //   if (a.rank > b.rank) {
  //     return 1;
  //   }
  //   if (a.rank < b.rank) {
  //     return -1;
  //   }
  //   return 0;
  // });

  filteredTasks.forEach((task) => {
    switch (task.category) {
      case 'A':
        categorizedFilteredTasks.TasksCatA.push(
          <Task
            className="task task--red"
            key={task._id}
            task={task}
          />,
        );
        break;

      case 'B':
        categorizedFilteredTasks.TasksCatB.push(
          <Task
            className="task task--orange"
            key={task._id}
            task={task}
          />,
        );
        break;

      case 'C':
        categorizedFilteredTasks.TasksCatC.push(
          <Task
            className="task task--green"
            key={task._id}
            task={task}
          />,
        );
        break;

      case 'D':
        categorizedFilteredTasks.TasksCatD.push(
          <Task
            className="task task--lightgray"
            key={task._id}
            task={task}
          />,
        );
        break;

      default:
        console.log(`There seems to be no category assigned to the following task: ${task}`);
    }
  });
  return categorizedFilteredTasks;
};

// tslint:disable-next-line: variable-name
const Main: React.FC<MainProps> = ({ filteredTasks }) => {
  const categorizedFilteredTasks = categorizeFilteredTasks(filteredTasks);

  return (
    <main className="main" id="main">
      <div className="main__column-category main__column-A">
        <span
          className="main__column-header"
        >
          Important and Urgent ({categorizedFilteredTasks.TasksCatA.length})
        </span>
        {categorizedFilteredTasks.TasksCatA}
      </div>
      <div className="main__column-category main__column-B">
        <span
          className="main__column-header"
        >
          Important ({categorizedFilteredTasks.TasksCatB.length})
        </span>
        {categorizedFilteredTasks.TasksCatB}
      </div>
      <div className="main__column-category main__column-C">
        <span
          className="main__column-header">
          Urgent ({categorizedFilteredTasks.TasksCatC.length})
        </span>
        {categorizedFilteredTasks.TasksCatC}
      </div>
      <div className="main__column-category main__column-D">
        <span
          className="main__column-header"
        >
          Not Important and not Urgent ({categorizedFilteredTasks.TasksCatD.length})
        </span>
        {categorizedFilteredTasks.TasksCatD}
      </div>
    </main>
  );
};

// Main.propTypes = {
//   filteredTasks: PropTypes.arrayOf(PropTypes.object).isRequired,
// };

export default Main;
