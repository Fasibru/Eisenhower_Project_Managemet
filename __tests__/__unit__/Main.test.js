import React from 'react';
import { shallow } from 'enzyme';
import Main from '../../src/client/components/js/Main';

describe('Testing <Main />', () => {
  const fnMock = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Main
      filteredTasks={[{
        _id: '2',
        rank: -999,
        category: 'A',
        title: 'Task Title A',
        description: 'Task Description A',
        completed: false,
      },
      {
        _id: '21',
        rank: -999,
        category: 'A',
        title: 'Task Title 2A',
        description: 'Task Description 2A',
        completed: false,
      },
      {
        _id: '3',
        rank: -999,
        category: 'B',
        title: 'Task Title B',
        description: 'Task Description B',
        completed: true,
      },
      {
        _id: '4',
        rank: -999,
        category: 'C',
        title: 'Task Title C',
        description: 'Task Description C',
        completed: true,
      },
      {
        _id: '5',
        rank: -999,
        category: 'D',
        title: 'Task Title D',
        description: 'Task Description D',
        completed: true,
      },
      {
        _id: '6',
        rank: -999,
        category: 'E',
        title: 'Task Title Default Cat',
        description: 'Task Description Default Cat',
        completed: true,
      }]}
      populateEditTask={fnMock}
    />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should categorize tasks properly', () => {
    const divCatA = wrapper.find('div').at(0);
    // console.log(divCatA.props().children.length);
    expect(divCatA.props().children.length).toEqual(2);
  });
});
