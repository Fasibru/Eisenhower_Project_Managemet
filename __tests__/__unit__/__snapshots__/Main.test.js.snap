// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Main /> Renders correctly 1`] = `
<main
  className="main"
>
  <div
    className="column-category"
  >
    <Task
      className="task task-category-A"
      key="2"
      populateEditTask={[MockFunction]}
      task={
        Object {
          "_id": "2",
          "category": "A",
          "completed": false,
          "description": "Task Description A",
          "rank": -999,
          "title": "Task Title A",
        }
      }
    />
    <Task
      className="task task-category-A"
      key="21"
      populateEditTask={[MockFunction]}
      task={
        Object {
          "_id": "21",
          "category": "A",
          "completed": false,
          "description": "Task Description 2A",
          "rank": -999,
          "title": "Task Title 2A",
        }
      }
    />
  </div>
  <div
    className="column-category"
  >
    <Task
      className="task task-category-B"
      key="3"
      populateEditTask={[MockFunction]}
      task={
        Object {
          "_id": "3",
          "category": "B",
          "completed": true,
          "description": "Task Description B",
          "rank": -999,
          "title": "Task Title B",
        }
      }
    />
  </div>
  <div
    className="column-category"
  >
    <Task
      className="task task-category-C"
      key="4"
      populateEditTask={[MockFunction]}
      task={
        Object {
          "_id": "4",
          "category": "C",
          "completed": true,
          "description": "Task Description C",
          "rank": -999,
          "title": "Task Title C",
        }
      }
    />
  </div>
  <div
    className="column-category"
  >
    <Task
      className="task task-category-D"
      key="5"
      populateEditTask={[MockFunction]}
      task={
        Object {
          "_id": "5",
          "category": "D",
          "completed": true,
          "description": "Task Description D",
          "rank": -999,
          "title": "Task Title D",
        }
      }
    />
  </div>
</main>
`;
